dask_kubernetes/operator/controller/tests/test_controller.py

```diff
@@ -374,6 +374,46 @@ async def test_recreate_scheduler_pod(k8s_cluster, kopf_runner, gen_cluster):
             )
 
 
+@pytest.mark.asyncio
+async def test_recreate_worker_pods(k8s_cluster, kopf_runner, gen_cluster):
+    with kopf_runner as runner:
+        async with gen_cluster() as (cluster_name, ns):
+            scheduler_deployment_name = "simple-scheduler"
+            worker_deployment_name = "simple-default-worker"
+            service_name = "simple-scheduler"
+            while scheduler_deployment_name not in k8s_cluster.kubectl(
+                "get", "pods", "-n", ns
+            ):
+                await asyncio.sleep(0.1)
+            while service_name not in k8s_cluster.kubectl("get", "svc", "-n", ns):
+                await asyncio.sleep(0.1)
+            while worker_deployment_name not in k8s_cluster.kubectl(
+                "get", "pods", "-n", ns
+            ):
+                await asyncio.sleep(0.1)
+            k8s_cluster.kubectl(
+                "delete",
+                "pods",
+                "-l",
+                "dask.org/cluster-name=simple,dask.org/component=worker",
+                "-n",
+                ns,
+            )
+            k8s_cluster.kubectl(
+                "wait",
+                "--for=condition=Ready",
+                "-l",
+                "dask.org/cluster-name=simple,dask.org/component=worker",
+                "pod",
+                "-n",
+                ns,
+                "--timeout=60s",
+            )
+            assert worker_deployment_name in k8s_cluster.kubectl(
+                "get", "pods", "-n", ns
+            )
```

Comment:
I spent some time tinkering with this test. I wrote this implementation which uses `kr8s` and tests a little more thoroughly that the Pods get created, become Ready, get deleted, get recreated and get back to Ready.

This has also inspired a few tweaks to kr8s so I may modify this a little further after it is merged.

```suggestion
    api = await kr8s.asyncio.api()
    with kopf_runner as runner:
        async with gen_cluster() as (cluster_name, ns):
            # Wait for worker Pods to be created
            while True:
                pods = await api.get(
                    "pods",
                    namespace=ns,
                    label_selector=f"dask.org/cluster-name={cluster_name},dask.org/component=worker",
                )
                if not pods:
                    await asyncio.sleep(0.1)
                    continue
                break

            # Store number of workers
            n_pods = len(pods)

            # Wait for worker Pods to be ready
            await asyncio.gather(
                *[pod.wait(conditions="condition=Ready", timeout=60) for pod in pods]
            )

            # Delete a worker Pod
            await pods[0].delete()

            # Wait for Pods to be recreated
            while True:
                pods = await api.get(
                    "pods",
                    namespace=ns,
                    label_selector=f"dask.org/cluster-name={cluster_name},dask.org/component=worker",
                )
                if len(pods) < n_pods:
                    await asyncio.sleep(0.1)
                    continue
                break

            # Wait for worker Pods to be ready
            await asyncio.gather(
                *[pod.wait(conditions="condition=Ready", timeout=60) for pod in pods]
            )
```

Reply:
I think I'm going to merge this and then update things in a follow up.

Reply:
Follow up is https://github.com/dask/dask-kubernetes/pull/743